clippy_lints/src/utils/internal_lints.rs

```diff
@@ -142,6 +145,64 @@ declare_clippy_lint! {
     "found 'default lint description' in a lint declaration"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** When `span_lint_and_then` function call and its
+    /// closure have one statment that is a method call of `span_suggestion`,
+    /// `span_help`, or `span_note` are using the same span argument,
+    /// or when closure have one statment that is a function call of `note` or
+    /// `help`, they should be replaced with one of three function calls which is
+    /// `span_lint_and_sugg`, span_lint_and_help` or `span_lint_and_note`.
+    ///
+    /// **Why is this bad?**
```

Comment:
I am not very clear why this is bad.

Reply:
We have wrapper functions for this, so we should use them wherever possible.

Reply:
Using wrapper functions mades intention more clear?

Reply:
I have thought about my question, and In fact what I want to ask should be phrased is why we want to make a wrapping function. What is the benefit to make it? I have read some refactor book before, and I am often confuse why coding in that special way improves the code.

Reply:
Ah ok, I'll split my answer to your question:

1. Why is this bad in the sense of the lint?

The lint message and the help/suggestion message have the same span most of the time. So Clippy has (wrapper) functions for this case. These wrappers make it easier to produce a lint+help message, since no closure is required, but you just have to pass some strings.

2. Why should you use wrapper functions at all?

The answer is convenience (and readability). Let's look at a simple (imaginary) case:

```rust
fn do_something(x: u32) -> Result<u32, Error> { ... }
```

If that is a function, that you use really often in your code, where the `x` argument is `0` (but sometimes is something else), you may want to define a function

```rust
fn do_something_zero() -> Result<u32, Error> { do_something(0) }
```

so that you don't have to type out the `0` argument.

If we transfer this to our `span_lint_and_then` function, the `|db| { ... }` argument is most of the time the same:

```rust
|db| {
    db.span_suggestion(span, msg, sugg, appl)
}
```

So we have a wrapper function, so you only have to pass `msg` and `sugg` and the wrapper function takes care of calling `span_lint_and_then` with the correct closure. -> More convenient, more readable, more error resistent (you don't have to write a closure, but just strings)

Reply:
Thank you, now I see the matter clearer.